genapi/src/ivalue.rs

```diff
@@ -132,55 +134,228 @@ macro_rules! impl_ivalue_for_vid {
+        &self,
+        value: String,
+        _: &mut impl Device,
+        _: &impl NodeStore,
+        cx: &mut ValueCtxt<U, S>,
+    ) -> GenApiResult<()> {
+        cx.value_store_mut().update(*self, value);
+        Ok(())
+    }
 
-            fn set_value<U: ValueStore, S: CacheStore>(
-                &self,
-                value: $ty,
-                device: &mut impl Device,
-                store: &impl NodeStore,
-                cx: &mut ValueCtxt<U, S>,
-            ) -> GenApiResult<()> {
-                self.$expect_kind(store)?
-                    .set_value(value, device, store, cx)
-            }
+    fn is_readable<U: ValueStore, S: CacheStore>(
+        &self,
+        _: &mut impl Device,
+        _: &impl NodeStore,
+        _: &mut ValueCtxt<U, S>,
+    ) -> GenApiResult<bool> {
+        Ok(true)
+    }
 
-            fn is_readable<U: ValueStore, S: CacheStore>(
-                &self,
-                device: &mut impl Device,
-                store: &impl NodeStore,
-                cx: &mut ValueCtxt<U, S>,
-            ) -> GenApiResult<bool> {
-                self.$expect_kind(store)?.is_readable(device, store, cx)
-            }
+    fn is_writable<U: ValueStore, S: CacheStore>(
+        &self,
+        _: &mut impl Device,
+        _: &impl NodeStore,
+        _: &mut ValueCtxt<U, S>,
+    ) -> GenApiResult<bool> {
+        Ok(true)
+    }
+}
 
-            fn is_writable<U: ValueStore, S: CacheStore>(
-                &self,
-                device: &mut impl Device,
-                store: &impl NodeStore,
-                cx: &mut ValueCtxt<U, S>,
-            ) -> GenApiResult<bool> {
-                self.$expect_kind(store)?.is_writable(device, store, cx)
-            }
+impl IValue<i64> for NodeId {
+    fn value<U: ValueStore, S: CacheStore>(
+        &self,
+        device: &mut impl Device,
+        store: &impl NodeStore,
+        cx: &mut ValueCtxt<U, S>,
+    ) -> GenApiResult<i64> {
+        if let Some(i) = self.as_iinteger_kind(store) {
+            i.value(device, store, cx)
+        } else if let Some(f) = self.as_ifloat_kind(store) {
+            f.value(device, store, cx).map(|f| f as i64)
+        } else if let Some(e) = self.as_ienumeration_kind(store) {
+            e.current_value(device, store, cx)
+        } else {
+            Err(GenApiError::invalid_node(
+                format!("Node {:?} is not an integer, float, or enumeration", self).into(),
+            ))
         }
-    };
+    }
+
+    fn set_value<U: ValueStore, S: CacheStore>(
+        &self,
+        value: i64,
+        device: &mut impl Device,
+        store: &impl NodeStore,
+        cx: &mut ValueCtxt<U, S>,
+    ) -> GenApiResult<()> {
+        if let Some(i) = self.as_iinteger_kind(store) {
+            i.set_value(value, device, store, cx)
+        } else if let Some(f) = self.as_ifloat_kind(store) {
+            f.set_value(value as f64, device, store, cx)
+        } else {
```

Comment:
I felt it'd be a little dangerous to set an arbitrary value to an enum node. But, since `IValue` is just used internally to cast a type by following the given GenICam XML, I changed it to allow setting an arbitrary value to an enum node.